wp/lib/bap_wp/src/precondition.ml

```diff
@@ -232,88 +233,30 @@ let exp_to_z3 (exp : Exp.t) (env : Env.t) : Constr.z3_expr * Constr.t list * Con
-            | Call call -> begin
-                match Call.return call with
-                | Some _ -> None
-                | None ->
-                  match Call.target call with
-                  | Direct _ -> None
-                  | Indirect label ->
-                    let label_vars = Exp.free_vars label in
-                    match Var.Set.choose label_vars with
-                    | None -> None
-                    | Some target ->
-                      (* In BIR, the return address is stored on the stack and then
-                         saved to a virtual variable to pass into the call instruction. *)
-                      if Var.Set.length label_vars = 1 && Var.is_virtual target then
-                        Some target
-                      else
-                        None
-              end
-            | _ -> None)
-      in
-      (* Check to see if the target value was obtained from the stack,
-         i.e. #x := mem[RSP, el]:u64 *)
-      match ret_addr with
-      | None -> false
-      | Some addr ->
-        let defs = Term.to_sequence ~rev:true def_t blk in
-        Seq.exists defs ~f:(fun def ->
-            if Var.equal (Def.lhs def) addr then begin
-              match Def.rhs def with
-              | Load (_, Var addr, _, _) -> Target.CPU.is_sp addr
-              | _ -> false
-            end else
-              false))
-  in
-  let sp = Bil.var Target.CPU.sp in
-  match ret_block with
-  | None ->
-    debug "Sub %s has no return" (Sub.name sub);
-    sp
-  | Some blk ->
-    let defs = Term.to_sequence def_t blk in
-    Seq.fold defs ~init:sp ~f:(fun offset def ->
-        (* Looks for instructions in the form SP := SP +/- offset *)
-        if Var.equal (Def.lhs def) Target.CPU.sp then begin
-          match Def.rhs def with
-          | BinOp (PLUS, Var v, w) ->
-            if Var.equal v Target.CPU.sp then
-              BinOp (PLUS, offset, w)
-            else offset
-          | BinOp (MINUS, Var v, w) ->
-            if Var.equal v Target.CPU.sp then begin
-              warning "Stack pointer in %s is being decremented by %s before return"
-                (Sub.name sub) (Exp.to_string w);
-              BinOp (MINUS, offset, w)
-            end else offset
-          | _ -> offset
-        end else
-          offset)
-
-let increment_stack_ptr (post : Constr.t) (env : Env.t) (offset : Exp.t)
-  : Constr.t * Env.t =
+let increment_stack_ptr (post : Constr.t) (env : Env.t) : Constr.t * Env.t =
+  let ctx = Env.get_context env in
   let arch = Env.get_arch env in
   let module Target = (val target_of_arch arch) in
-  let sp, _ = Env.get_var env Target.CPU.sp in
-  let z3_off, _, _, env = exp_to_z3 offset env in
+  let sp, env = Env.get_var env Target.CPU.sp in
+  let width = Target.CPU.sp |> Var.typ |> typ_size in
+  (* TODO: We are currently incrementing the stack pointer by 8 for the return address.
+     We will have to change that offset depending on the archtecture. *)
```

Comment:
It actually ABI dependent but, fortunately, mostly relative to x86 family. Basically, if you will assume a C ABI, then you should decrement it by 4 for x86 and by 8 for x86_64. For all other architectures supported by BAP (and known to ABIs) there is no need for this correction.

Therefore, I'd suggest the following:
1) increment only if the architecture is `#x86` 
2) increment by 4 or 8 depending on arch (I would suggest: `Size.in_bytes (Arch.addr_size arch)`)

Then it will work fine for all architectures in the nearest future :)

Reply:
see also a realtive PR for Primus
https://github.com/BinaryAnalysisPlatform/bap/pull/955/files

Reply:
I see, interesting! There does seem to be some effort duplication across WP and Primus, but it's hard for me to see if we could easily resolve this (probably not).

Reply:
The main effort here is basically researching this problem, which is basically contained in the PR message, which is worthwhile to read :) Another commonality is that any analysis requires some kind of stubs and one of the reasons why many people started to use Primus is because how easy it is to stub things with Primus Lisp. That's why we decided to generalize Primus Lisp, and it will now become BAP Lisp. The plan was to release it together with BAP 2.0 but I didn't have time to do this, but will return to it as soon as will get it released (basically in a couple of weeks). We will soon be able to use BAP Lisp to specify facts for BAP's knowledge base. The first example that comes to mind (other than the already existing stubs) is writing lifters. And I believe that WP can also benefit from stubs written in BAP Lisp.  

But back to this PR. There is one caveat, and I can't infer immediately from the PR, whether it is handled correctly. The stack pointer is corrected by the subroutine code (when the `ret` instruction is executed). Therefore if the subroutine code is present, there should be no corrections (otherwise we will overcorrect). So we should be careful and do corrections only for stubs (and for x86).

Reply:
Right, we are aware of the issue, and we'll fix this in a cleaner way at some point. 

For now things happen to work however, because of how we treat inlined code: basically the call becomes a `nop` and the return as well, because of a silly bug (we treat indirect jumps as `nop`s).

Reply:
Thanks for taking a look at this Ivan! I was unaware that the correction was only needed in x86. Right now, we're only correcting the stack pointer when we are using our stubs to handle the targets of the function calls.